### PR TITLE
SAM debug: more test coverage + edge cases

### PR DESCRIPTION
**Pull request by justinmk3**
- add direct coverage of `debugConfiguration.getCodeRoot()`, `debugConfiguration.getHandlerName()`
- test python 3.7
- test invalid template.yaml

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/test/shared/sam/debugger/samDebugConfigProvider.test.ts [hunk 1]
```diff
@@ -103,6 +108,11 @@ describe('AwsSamDebugConfigurationProvider', async () => {
             assert.deepStrictEqual(await debugConfigProvider.provideDebugConfigurations(undefined), [])
             // Workspace with no templates:
             assert.deepStrictEqual(await debugConfigProvider.provideDebugConfigurations(fakeWorkspaceFolder), [])
+
+            // Malformed template.yaml:
+            await strToYamlFile('bogus', tempFile.fsPath)
```
Comment (bryceitoc9) [inline src/test/shared/sam/debugger/samDebugConfigProvider.test.ts:113]: JS-YAML still parses this as valid YAML. Use https://github.com/aws/aws-toolkit-vscode/blob/ab650f54d0a4526aadcbff2979136f27ba735717/src/test/shared/cloudformation/cloudformationTestUtils.ts#L81 instead (this definitely doesn't parse)

File: src/test/shared/sam/debugger/samDebugConfigProvider.test.ts [hunk 2]
```diff
@@ -520,6 +530,82 @@ describe('AwsSamDebugConfigurationProvider', async () => {
...
+                },
+                lambda: {
+                    runtime: 'python3.7',
+                },
+            }
+            ;(c as any).__noInvoke = true
+            const actual = (await debugConfigProvider.resolveDebugConfiguration(
+                folder,
+                c
+            ))!! as DotNetCoreDebugConfiguration
```
Comment (bryceitoc9) [inline src/test/shared/sam/debugger/samDebugConfigProvider.test.ts:555]: I'm pretty sure trailing `!!` === trailing `!`

File: src/test/testUtil.ts [hunk 3]
```diff
@@ -27,6 +27,6 @@ export function getWorkspaceFolder(dir: string): vscode.WorkspaceFolder {
 /**
  * Asserts that filepaths are equal, after normalizing for platform differences.
  */
-export function assertEqualPaths(actual: string, expected: string) {
-    assert.strictEqual(pathutil.normalize(actual), pathutil.normalize(expected))
+export function assertEqualPaths(actual: string, expected: string, message?: string | Error) {
```
Comment (bryceitoc9) [inline src/test/testUtil.ts:30]: Any reason why this was added to this change?

File: src/test/shared/sam/debugger/samDebugConfigProvider.test.ts [hunk 4]
```diff
@@ -520,6 +530,82 @@ describe('AwsSamDebugConfigurationProvider', async () => {
...
+                invokeTarget: {
+                    target: CODE_TARGET_TYPE,
+                    lambdaHandler: 'app.lambda_handler',
+                    projectRoot: 'hello_world',
+                },
+                lambda: {
+                    runtime: 'python3.7',
+                },
+            }
+            ;(c as any).__noInvoke = true
```
Comment (bryceitoc9) [inline src/test/shared/sam/debugger/samDebugConfigProvider.test.ts:551]: What's the point of this?

File: src/test/shared/sam/debugger/samDebugConfigProvider.test.ts (hunk 1 above)
Comment (justinmk3) [reply src/test/shared/sam/debugger/samDebugConfigProvider.test.ts:113]: Thanks. We can have more insightful tests by checking the error popups or log messages explicitly. Will add a tracking ticket for that.

File: src/test/testUtil.ts (hunk 3 above)
Comment (justinmk3) [reply src/test/testUtil.ts:30]: No reason.

File: src/test/shared/sam/debugger/samDebugConfigProvider.test.ts (hunk 4 above)
Comment (justinmk3) [reply src/test/shared/sam/debugger/samDebugConfigProvider.test.ts:551]: It is temporary, but there is a note here https://github.com/aws/aws-toolkit-vscode/blob/7f77a292f7ddefba28a3cd2c67c4f2cf0ec23337/src/shared/sam/debugger/awsSamDebugger.ts#L117

File: src/test/shared/sam/debugger/samDebugConfigProvider.test.ts (hunk 1 above)
Comment (justinmk3) [reply src/test/shared/sam/debugger/samDebugConfigProvider.test.ts:113]: I made that change, but then reverted it: I am not looking for unparsable YAML, but just a bad template.yaml. I.e. I want a semantic error, not a syntax error (`YAMLException`).

File: src/test/shared/sam/debugger/samDebugConfigProvider.test.ts [hunk 5]
```diff
@@ -559,7 +559,7 @@ describe('AwsSamDebugConfigurationProvider', async () => {
                 runtime: 'python3.7',
                 runtimeFamily: lambdaModel.RuntimeFamily.Python,
                 type: 'python', // Input "aws-sam", output "python".
-                handlerName: '___vsctk___debug.lambda_handler',
+                handlerName: 'app___vsctk___debug.lambda_handler',
```
Comment (justinmk3) [inline src/test/shared/sam/debugger/samDebugConfigProvider.test.ts:562]: adjusted after https://github.com/aws/aws-toolkit-vscode/pull/1056

**Comment by justinmk3** [comment]
artifact: [aws-toolkit-vscode-1.9.0-SNAPSHOT.vsix.zip](https://github.com/aws/aws-toolkit-vscode/files/4560383/aws-toolkit-vscode-1.9.0-SNAPSHOT.vsix.zip)
